misc/speedtest.py

```diff
@@ -0,0 +1,300 @@
+        if reporturl is None:
+            reporturl = pageurl
+        logger.debug('GET %s', pageurl)
+        response = urllib.request.urlopen(pageurl)
+        if (response.code != 200) or not response.getheader('Content-Type').startswith('application/json'):
+            raise Exception("bad response to {!r}: {}".format(pageurl, response.reason))
+        ob = json.loads(response.read())
+        txns = ob.get('transactions', [])
+        txcount += len(txns)
+        count += 1
+        dt = time.time() - start
+        if (n and (count >= n)) and (not minTime or (dt > minTime)):
+            break
+        if not n and minTime and (dt > minTime):
+            break
+        if maxTime and (dt > maxTime):
+            break
+    logger.info('asset txns: %d assets in %0.2f seconds, %.1f /s; %d txns, %.1f txn/s', count, dt, count/dt, txcount, txcount/dt)
+    return {'url':reporturl, 'assets': count, 'seconds': dt, 'txns': txcount}
+
+# /v2/assets/{asset-id}/balances -- maybe add index to account_asset table?
+#
+# To make fast:
+# CREATE INDEX CONCURRENTLY IF NOT EXISTS account_asset_asset ON account_asset (assetid, addr ASC);
+def assetBalances(rooturl, assets, n=1000, minTime=None, maxTime=10, ntxns=1000):
+    rootparts = urllib.parse.urlparse(rooturl)
+    rawurl = list(rootparts)
+    atxnsurl = list(rootparts)
+    assets = list(assets.keys())
+    random.shuffle(assets)
+    count = 0
+    balcount = 0
+    start = time.time()
+    reporturl = None
+    for assetid in assets:
+        atxnsurl[2] = os.path.join(rawurl[2], 'v2', 'assets', str(assetid), 'balances')
+        query = {'limit':ntxns}
+        atxnsurl[4] = urllib.parse.urlencode(query)
+        pageurl = urllib.parse.urlunparse(atxnsurl)
+        if reporturl is None:
+            reporturl = pageurl
+        logger.debug('GET %s', pageurl)
+        response = urllib.request.urlopen(pageurl)
+        if (response.code != 200) or not response.getheader('Content-Type').startswith('application/json'):
+            raise Exception("bad response to {!r}: {}".format(pageurl, response.reason))
+        ob = json.loads(response.read())
+        bals = ob.get('balances', [])
+        balcount += len(bals)
+        count += 1
+        dt = time.time() - start
+        if (n and (count >= n)) and (not minTime or (dt > minTime)):
+            break
+        if not n and minTime and (dt > minTime):
+            break
+        if maxTime and (dt > maxTime):
+            break
+    logger.info('asset balances: %d assets in %0.2f seconds, %.1f /s; %d bals, %.1f bal/s', count, dt, count/dt, balcount, balcount/dt)
+    return {'url':reporturl, 'assets': count, 'seconds': dt, 'balances': balcount}
+
+# TODO:
+# /v2/applications -- this should be fast because it's a single table scan
+# /v2/applications/N -- single row lookup, easy
+# /v2/assets/{asset-id} -- single row lookup, easy
+# /v2/blocks/{round-number} -- single row lookup, easy
+# /v2/transactions what search opts?
```

Comment:
These should be tested. 

For the transactions endpoint, we should mainly test with date and rounds ranges.

Reply:
The first four I'm pretty sure will be good enough. I'm pretty sure txn-by-round will be fast since that's the primary key. I guess a good txn-by-date test would get the first txn and last txn and pick random dates between getting the 1000 txn after that. I'd kinda like to get this work in as good-ish and do a follow-on ticket for another round of improvements?

Reply:
These are currently slow so we definitely need to cover them:
* asset-id/application-id (looks like you got that above)
* before-time/after-time

The other options:
* address-role
* currency GT/LT
* max/min round
* note-prefix
* rekey-to
* round
* sig-type
* tx-type
* txid

There are also a lot of account search options:
* application/asset
* auth-addr
* currency GT/LT
* at round

/v2/assets
* creator
* name
* unit

/v2/applications
* n/a - though creator should probably be added to match /v2/assets